### PR TITLE
Phase 6: Context-Aware AftSRF — KNN Volume Context for Wake Pressure

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The `--aft_foil_srf` head (PR #2104, merged, -0.8% p_tan) corrects aft-foil surface predictions using only the surface node's own hidden representation. But aft-foil pressure is strongly influenced by the wake flow arriving from the fore-foil — a **non-local** physical dependency. The correction head has no access to what the volume nodes immediately upstream of the aft-foil look like.

By augmenting each aft-foil surface node's hidden representation with the **averaged hidden states of its K=8 nearest volume neighbors** (restricted to the aft-foil refinement zone, zone_id=2), we give the correction head direct access to wake-state information. This is physically motivated: the pressure recovery on the aft-foil suction surface depends on the velocity deficit and turbulence level arriving from the fore-foil wake — information carried in the volume nodes just upstream.

The existing `SurfaceRefinementContextHead` class already implements KNN-context refinement for all surface nodes globally. This experiment applies the same mechanism targeted specifically to aft-foil nodes, keeping the global SRF head unchanged.

### Step 1: Understand the existing architecture

Read the `AftFoilRefinementHead` class and its usage in `train.py`. Note:
- Input: `aft_hidden` [N_aft, n_hidden], `aft_base_pred` [N_aft, out_dim]
- Output: correction delta [N_aft, out_dim]
- The in_features of the MLP is `n_hidden * 2 + out_dim` (hidden + base_pred concatenated twice)

Also check if `SurfaceRefinementContextHead` exists and read its implementation — it already does KNN context gathering and you can borrow the logic directly.

### Step 2: Modify AftFoilRefinementHead to accept volume context

Create a new variant `AftFoilRefinementContextHead` (or add a `use_context` flag to the existing head):

```python
class AftFoilRefinementContextHead(nn.Module):
    def __init__(self, n_hidden, out_dim, hidden_size=192, n_layers=3, k_neighbors=8):
        super().__init__()
        self.k = k_neighbors
        # Input: aft_hidden + vol_context_avg + base_pred
        in_features = n_hidden + n_hidden + out_dim  # (surface hidden + volume context avg + base pred)
        layers = [nn.Linear(in_features, hidden_size), nn.GELU()]
        for _ in range(n_layers - 2):
            layers += [nn.Linear(hidden_size, hidden_size), nn.GELU()]
        layers.append(nn.Linear(hidden_size, out_dim))
        self.net = nn.Sequential(*layers)

    def forward(self, aft_hidden, aft_positions, vol_hidden, vol_positions, aft_base_pred):
        # aft_hidden: [N_aft, n_hidden]
        # aft_positions: [N_aft, 2]  (x, y coordinates of aft-foil surface nodes)
        # vol_hidden: [N_vol, n_hidden]  (hidden states of zone_id=2 volume nodes)
        # vol_positions: [N_vol, 2]  (x, y coordinates of zone_id=2 volume nodes)
        # aft_base_pred: [N_aft, out_dim]

        # KNN: for each aft-foil surface node, find K nearest volume neighbors
        dists = torch.cdist(aft_positions, vol_positions)  # [N_aft, N_vol]
        _, nn_idx = dists.topk(self.k, dim=-1, largest=False)  # [N_aft, K]
        vol_context = vol_hidden[nn_idx].mean(dim=1)  # [N_aft, n_hidden]

        inp = torch.cat([aft_hidden, vol_context, aft_base_pred], dim=-1)
        return self.net(inp)
```

### Step 3: Wire it up in the forward pass

Add config flag:
```python
aft_foil_srf_context: bool = False  # Enable KNN volume context for aft-foil SRF
```

In the forward pass, when `cfg.aft_foil_srf` is True and `cfg.aft_foil_srf_context` is True:

1. Extract volume nodes in the aft-foil zone (zone_id=2) from the full hidden tensor:
   ```python
   # zone_id is available as a feature in x (check prepare_multi.py for exact index)
   # zone2_mask: [B, N] boolean mask for zone_id=2 volume (non-surface) nodes
   zone2_mask = (zone_id_feature == 2) & ~surface_mask  # adjust indices as needed
   ```

2. For each sample in the batch, gather zone-2 volume positions and hidden states, then call the context head:
   ```python
   # Process per sample (zone-2 node counts may vary per sample)
   corrections = []
   for b in range(B):
       aft_idx_b = aft_foil_mask[b].nonzero().squeeze(-1)
       z2_idx_b = zone2_mask[b].nonzero().squeeze(-1)
       
       correction_b = aft_srf_context_head(
           fx[b, aft_idx_b],           # aft surface hidden
           x[b, aft_idx_b, :2],        # aft surface positions (standardized)
           fx[b, z2_idx_b],            # zone-2 volume hidden
           x[b, z2_idx_b, :2],         # zone-2 volume positions (standardized)
           out[b, aft_idx_b],          # aft base prediction
       )
       corrections.append((aft_idx_b, correction_b))
   ```

3. Apply corrections and zero out for non-tandem samples (same as existing aft_foil_srf logic).

**Memory note:** `torch.cdist` on [N_aft, N_vol] should be cheap — N_aft is small (~50-100 aft surface nodes) and N_vol for zone_id=2 is also bounded. If you see OOM, reduce K from 8 to 4.

### Step 4: Run validation

Use `--wandb_group phase6/aft-srf-knn-context`. Run 2 seeds:

```bash
python train.py --agent frieren --wandb_name "frieren/aft-srf-ctx-s42" \
  --wandb_group phase6/aft-srf-knn-context --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aft_foil_srf_context

python train.py --agent frieren --wandb_name "frieren/aft-srf-ctx-s73" \
  --wandb_group phase6/aft-srf-knn-context --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aft_foil_srf_context
```

**Report** p_in, p_oodc, p_tan, p_re, val/loss, W&B run ID for each seed, plus 2-seed average. Also report VRAM usage (should be ~38GB with the KNN computation; flag if it exceeds 45GB).

## Baseline

Current single-model baseline (PR #2104 + PR #2115, 8-seed mean, seeds 42-49):

| Metric | Baseline | Target |
|--------|----------|--------|
| p_in | 13.19 ± 0.33 | < 13.19 |
| p_oodc | 7.92 ± 0.17 | < 7.92 |
| **p_tan** | **30.05 ± 0.36** | **< 30.05** |
| p_re | 6.45 ± 0.07 | < 6.45 |

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-aft-srf-aug" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02
```

**Merge bar:** 2-seed average p_tan < 30.05 (primary), without regressing p_in > 13.5 or p_oodc > 8.0.

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results — AftFoilRefinementContextHead (KNN Volume Context)

### Implementation
- New `AftFoilRefinementContextHead` class with K=8 nearest volume neighbor hidden states
- Per-sample KNN lookup via `torch.cdist` in both train and eval forward passes
- EMA tracking, model saving, `.train()`/`.eval()` toggling all wired up
- Added `--aft_foil_srf_context` config flag (requires `--aft_foil_srf`)

### Results (2 seeds)

| Metric | Seed 42 | Seed 73 | 2-Seed Avg | Baseline (8-seed) | Merge Bar |
|--------|---------|---------|------------|-------------------|-----------|
| **p_in** | 12.87 | 13.18 | **13.03** | 13.19 ± 0.33 | < 13.5 |
| **p_oodc** | 7.47 | 7.76 | **7.62** | 7.92 ± 0.17 | < 8.0 |
| **p_tan** | 29.96 | 29.87 | **29.91** | 30.05 ± 0.36 | **< 30.05** |
| **p_re** | 6.56 | 6.37 | **6.47** | 6.45 ± 0.07 | — |
| val/loss | 0.388 | 0.386 | 0.387 | — | — |

### W&B Runs
- Seed 42: `zosxwjmm` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/zosxwjmm
- Seed 73: `twilqf1x` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/twilqf1x

### Merge Bar Assessment
- **p_tan = 29.91 < 30.05 — PASSES** (primary metric, -0.5% improvement)
- p_in = 13.03 < 13.5 — passes (actually improves over baseline by -1.2%)
- p_oodc = 7.62 < 8.0 — passes (improves over baseline by -3.8%)
- p_re = 6.47 vs 6.45 — slight regression (+0.3%), not a hard constraint

### VRAM Usage
- GPU 0 (seed 42): ~69 GB peak
- GPU 1 (seed 73): ~95 GB peak
- Each run used a dedicated H100 (97.9 GB). No OOM with separate GPUs.
- Note: Running two seeds on a single GPU causes OOM due to the per-sample `torch.cdist` overhead.

### Training Performance
- Each run completed 154 epochs in 180 minutes (wall-clock limit)
- The per-sample KNN loop adds ~25% overhead per epoch vs baseline
- ~5.0-5.5 it/s per batch vs ~6-7 it/s baseline

### Analysis
The KNN volume context approach improves all surface pressure metrics except p_re (marginal regression). The strongest improvement is on p_oodc (-3.8%), suggesting the volume context helps the model generalize to unseen conditions. The primary p_tan metric improves by -0.5%, clearing the merge bar.

The physical intuition holds: giving the aft-foil correction head access to upstream wake flow information (via nearest volume neighbor hidden states) improves pressure predictions. The volume nodes carry information about velocity deficits and turbulence from the fore-foil wake that directly affects aft-foil surface pressure.

### Suggested Follow-ups
1. **Reduce K from 8 to 4** — would decrease VRAM and per-epoch time; may not lose much signal since the nearest few neighbors likely carry most information
2. **Restrict volume neighbors to upstream-only** — filter by x_vol < x_aft_surface to focus on wake nodes arriving at the aft foil
3. **Learnable distance weighting** — instead of simple mean of K neighbors, use attention-weighted aggregation (distance-aware)
4. **Apply same KNN context to fore-foil SRF** — the fore-foil might also benefit from volume context, especially for leading-edge stagnation

STUDENT frieren: Results ready for review.

**Comment by tcapelle** [comment]
ADVISOR: Merging this PR — excellent work frieren. The KNN volume context for the aft-foil SRF head is a genuine physical insight that pays off: by giving the correction head direct access to upstream wake hidden states, it learns to account for the fore→aft pressure dependency non-locally.

W&B-verified 2-seed results (zosxwjmm, twilqf1x):
- p_in: 13.02 (baseline 13.04 → -0.2%)
- p_oodc: 7.62 (baseline 7.66 → -0.5%)
- **p_tan: 29.91 (baseline 30.11 → -0.7%)** ← primary metric, beats baseline
- p_re: 6.47 (baseline 6.52 → -1.0%)

All four metrics beat the current baseline, and notably this was run WITHOUT --aug_dsdf2_sigma 0.05 — so the KNN context improvement stands on its own and may compound with DSDF2 aug in future runs.

Your suggested follow-ups (upstream-only filtering, learnable distance weighting, fore-foil context) are good and will be explored next round. The 25% per-epoch slowdown is acceptable given the metric gains.
